src/main/scala/passes/HoistSlowBinop.scala

```diff
@@ -0,0 +1,122 @@
+package fuselang.passes
+
+import scala.{PartialFunction => PF}
+import fuselang.common._
+import Syntax._
+// import CodeGenHelpers._
+// import CompilerError._
+import Transformer._
+import EnvHelpers._
+
+object HoistSlowBinop extends TypedPartialTransformer {
+
+  case class ExprEnv(map: Map[Expr, CLet])
+      extends ScopeManager[ExprEnv]
+      with Tracker[Expr, CLet, ExprEnv] {
+    def merge(that: ExprEnv) = {
+      ExprEnv(this.map ++ that.map)
+    }
+    def get(key: Expr) = this.map.get(key)
+    def add(key: Expr, value: CLet) = {
+      ExprEnv(this.map + (key -> value))
+    }
+  }
+
+  type Env = ExprEnv
+  val emptyEnv = ExprEnv(Map())
+
+  private val slowBinops = List("*", "/", "%")
+
+  var idx: Map[String, Int] = Map();
+  def genName(base: String): Id = {
+    // update idx
+    idx.get(base) match {
+      case Some(n) => idx = idx + (base -> (n + 1))
+      case None => idx = idx + (base -> 0)
+    }
+    Id(s"$base${idx(base)}_")
+  }
+
+  def construct(
+      cmd: Command,
+      env: Env,
+      acc: Command = CEmpty
+  ): (Command, Env) = {
+    if (env.map.values.isEmpty && acc == CEmpty) {
+      cmd -> emptyEnv
+    } else {
+      CSeq.smart(env.map.values.toSeq :+ acc :+ cmd) -> emptyEnv
+    }
+  }
+
+  def binopRecur(expr: Expr, env: Env): (Expr, Env) = {
+    expr match {
+      // only recur when children are binops
+      case EBinop(_, _, _) => rewriteE(expr)(env)
+      case _ => (expr, env)
+    }
+  }
+
+  def myRewriteE: PF[(Expr, Env), (Expr, Env)] = {
+    case (e @ EBinop(op, left, right), env) if slowBinops.contains(op.op) => {
```

Comment:
A large part of this code overlaps with the HoistMemoryReads traversal (unsurprisingly). Is there a way to abstract this pattern into a class that can be instantiated with a `Expr => boolean` function and hoists everything that it gets `true` for?